openapitor/src/types/mod.rs

```diff
@@ -96,34 +96,73 @@ pub fn generate_types(spec: &openapiv3::OpenAPI, opts: crate::Opts) -> Result<Ty
         opts,
     };
 
-    if let Some(components) = &spec.components {
-        // Parse the schemas.
-        for (name, schema) in &components.schemas {
-            // Let's get the schema from the reference.
-            let schema = schema.get_schema_from_reference(spec, true)?;
-            // Let's handle all the kinds of schemas.
-            type_space.render_schema(name, &schema)?;
-        }
-        // Parse the parameters.
-        for (name, parameter) in &components.parameters {
-            let schema = (&parameter.expand(spec)?).data()?.format.schema()?;
-            // Let's get the schema from the reference.
-            let schema = schema.get_schema_from_reference(spec, true)?;
-            // Let's handle all the kinds of schemas.
-            type_space.render_schema(name, &schema)?;
-        }
+    // The 'components' field of an OpenAPI object stores
+    // definitions of various types/shapes of data.
+    // We will generate a Rust type for these.
+    // If there aren't any of them, then there's no work left to do!
+    // So we can return early.
+    let Some(components) = &spec.components else {
+        return Ok(type_space);
+    };
+
+    // Parse the schemas.
+    // Schemas are "definition of input and output data types",
+    // i.e. the shape of requests to and responses from endpoints.
+    //
+    // Schemae are either defined as:
+    //  - shared schemae that get used in many different places
+    //  - a description of some parameter used in a request
+    //    (e.g. a path/query/header parameter)
+    //  - a description of a request body
+    //  - a description of a response body
+    //
+    // We need to search each of these 4 places in the spec for schemae, and
+    // generate Rust types for them.
+
+    let mut schemas = Vec::new();
```

Comment:
previously we were exploring each place the schemae could be and rendering from each. Instead I want to explore each place, build up the list of schemae, and then render them all at once. Separate search and rendering rather than interleaving them.